pkg/tun/tun.go

```diff
@@ -0,0 +1,164 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package tun wraps the wg tunnel implementation and adds packet filters.
+package tun
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net/netip"
+	"os"
+
+	"golang.zx2c4.com/wireguard/tun"
+)
+
+// PacketHeader is the decoded representation of the wireguard packet.
+type PacketHeader struct {
+	SourceAddr      netip.Addr
+	DestinationAddr netip.Addr
+	FlowLabel       uint32
+	PayloadLength   uint16
+	Version         uint8
+	TrafficClass    uint8
+	NextHeader      uint8
+	HopLimit        uint8
+}
+
+// PacketFilter is the function for filtering the packets going through the wireguard tunnel.
+type PacketFilter func(PacketHeader) bool
```

Comment:
this is not a generic PacketFilter, but rather a PacketFilter for a specific flow, I guess it should be `InputPacketFilter` (?) as it filters packets received from Wireguard